zetaclient/evm/evm_signer.go

```diff
@@ -324,246 +332,233 @@ func (signer *Signer) SignCommandTx(
-				cctx.GetCurrentOutTxParam().Amount.BigInt(),
-				cctx.GetCurrentOutTxParam().OutboundTxTssNonce,
-				gasprice,
-				height,
-			)
+			logger.Info().Msgf("SignWithdrawTx: %d => %s, nonce %d, gasPrice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, txData.gasPrice)
+			tx, err = signer.SignWithdrawTx(&txData)
 		}
 		if cctx.GetCurrentOutTxParam().CoinType == common.CoinType_ERC20 {
-			asset := ethcommon.HexToAddress(cctx.InboundTxParams.Asset)
-			logger.Info().Msgf("SignERC20WithdrawTx: %d => %s, nonce %d, gasprice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, gasprice)
-			tx, err = signer.SignERC20WithdrawTx(
-				to,
-				asset,
-				cctx.GetCurrentOutTxParam().Amount.BigInt(),
-				gasLimit,
-				cctx.GetCurrentOutTxParam().OutboundTxTssNonce,
-				gasprice,
-				height,
-			)
+			logger.Info().Msgf("SignERC20WithdrawTx: %d => %s, nonce %d, gasPrice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, txData.gasPrice)
+			tx, err = signer.SignERC20WithdrawTx(&txData)
 		}
 	} else if cctx.CctxStatus.Status == types.CctxStatus_PendingRevert {
-		logger.Info().Msgf("SignRevertTx: %d => %s, nonce %d, gasprice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, gasprice)
-		tx, err = signer.SignRevertTx(
-			ethcommon.HexToAddress(cctx.InboundTxParams.Sender),
-			big.NewInt(cctx.OutboundTxParams[0].ReceiverChainId),
-			to.Bytes(),
-			big.NewInt(cctx.GetCurrentOutTxParam().ReceiverChainId),
-			cctx.GetCurrentOutTxParam().Amount.BigInt(),
-			gasLimit,
-			message,
-			sendhash,
-			cctx.GetCurrentOutTxParam().OutboundTxTssNonce,
-			gasprice,
-			height,
-		)
+		logger.Info().Msgf("SignRevertTx: %d => %s, nonce %d, gasPrice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, txData.gasPrice)
+		txData.srcChainID = big.NewInt(cctx.OutboundTxParams[0].ReceiverChainId)
+		txData.toChainID = big.NewInt(cctx.GetCurrentOutTxParam().ReceiverChainId)
+
+		tx, err = signer.SignRevertTx(&txData)
 	} else if cctx.CctxStatus.Status == types.CctxStatus_PendingOutbound {
-		logger.Info().Msgf("SignOutboundTx: %d => %s, nonce %d, gasprice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, gasprice)
-		tx, err = signer.SignOutboundTx(
-			ethcommon.HexToAddress(cctx.InboundTxParams.Sender),
-			big.NewInt(cctx.InboundTxParams.SenderChainId),
-			to,
-			cctx.GetCurrentOutTxParam().Amount.BigInt(),
-			gasLimit,
-			message,
-			sendhash,
-			cctx.GetCurrentOutTxParam().OutboundTxTssNonce,
-			gasprice,
-			height,
-		)
+		logger.Info().Msgf("SignOutboundTx: %d => %s, nonce %d, gasPrice %d", cctx.InboundTxParams.SenderChainId, toChain, cctx.GetCurrentOutTxParam().OutboundTxTssNonce, txData.gasPrice)
+		tx, err = signer.SignOutboundTx(&txData)
 	}
 
 	if err != nil {
```

Comment:
The else if ensures we only execute one code block, if we remove the else if's it will proceed to check all the other conditions and possibly sign the transaction again. The assumption above works only if the signature fails.

I can add the return err in each if block but like you said it will make the function longer. Either way I think there is a better solution, we can have a router of some kind to replace the else if's but it will take some time to think about the implementation.

Reply:
Not handing the error right after it is returned makes code maintainability harder. Having a longer function is less a concern. We can create new low-level methods to process the different cases.

Reply:
I'm going to just return the error for each condition since this PR is getting a bit long. I will open another PR to clean this part up some more.

Reply:
I don't get it. It's a trivial change. Why not adding it here?

Reply:
The change has been made already, check the code. If you read my comment I've added the error check in each case and I've opened another issue to fix some other things in this file.